src/BCFSerialize.cc

```diff
@@ -125,6 +157,41 @@ int bcf_shallow_compare(const bcf1_t *x, const bcf1_t *y) {
     return 1;
 }
 
+/* Adapted from [htslib::vcf.c::bcf_hdr_read] to read
+   from memory instead of disk.
+*/
+static Status bcf_raw_read_header(const char* buf,
+                                  int hdrlen,
+                                  int& consumed,
+                                  shared_ptr<bcf_hdr_t>& ans) {
+    if (strncmp(buf, "BCF\2\2", 5) != 0) {
+        return Status::Invalid("BCFSerialize::bcf_raw_read_header invalid BCF2 magic string");
+    }
+
+    ans = shared_ptr<bcf_hdr_t>(bcf_hdr_init("r"), &bcf_hdr_destroy);
+    int loc = 5;
+    int hlen;
+    memcpy(&hlen, &buf[loc], 4);
+    loc += 4;
+
+    // make sure we do not read beyond the buffer end
+    if (loc + hlen > hdrlen) {
+        return Status::Invalid("BCFSerialize::bcf_raw_read_header truncated header");
+    }
+
+    auto htxt = make_unique<char[]>(hlen);
+    memcpy(htxt.get(), &buf[loc], hlen);
+
+    // FIXME bcf_hdr_parse does not seem to check bounds -- potential security issue
```

Comment:
Since you're so familiar with these parts of the htslib code now, it'd be cool to write a boundschecking pull request to send upstream...not a high priority though